[CORL-1761] Show external moderation phases on the moderate card

nick-funk opened the pull request:
## What does this PR do?

Adds markers, "moderate by" state, and "automated actions" details to the comment moderation cards in the mod queues.

![image](https://user-images.githubusercontent.com/5751504/122790278-64423800-d275-11eb-98b5-08b0ea339088.png)

## What changes to the GraphQL/Database Schema does this PR introduce?

- Adds external moderation phase metadata to the comment's metadata

## How do I test this PR?

- Create an external moderation phase in Admin > Config > Moderation
  - Will require making a server to call an endpoint. I suggest using a quick express app with a POST endpoint to return some status like: 
    ```
      app.post("/mod", (req, res) => {
        res.send({
          actions: [
            {
              actionType: "FLAG",
              reason: "COMMENT_DETECTED_TOXIC",
            },
            {
              actionType: "FLAG",
              reason: "COMMENT_DETECTED_SPAM",
            },
          ],
        });
      });
    ```
- Create some comments as a non-admin/mod user
- See the comments get flagged with whatever you put into your mod phase and that the comments are marked with tags of the mod phases in the moderate card
- You can also force them into specific queues by making the mod result return `status: "REJECTED"` or similar

nick-funk committed:
Persist external mod phase actions to comment rev metadata

nick-funk committed:
Display external mod phase actions in mod card details

nick-funk committed:
Refactor external mod details status, flag UI

nick-funk committed:
Refactor out external mod history validation

nick-funk committed:
Shore up styling of external mod phase card details

nick-funk committed:
Allow "moderated by" on mod card to show external mod phase

nick-funk committed:
Better spacing for external mod phase markers

nick-funk committed:
Indicate if a phase was not run

nick-funk committed:
Update snapshots

nick-funk committed:
Move toxicity under mod card > details > automated actions

wyattjoh commented on src/core/server/models/comment/revision.ts:
```diff
@@ -45,6 +54,21 @@ export interface RevisionMetadata {
+    actions?: {
+      type?: ACTION_TYPE;
+      reason?: FLAG_REASON;
+    }[];
```
May I suggest nesting this one level deeper under `result` or something similar to distinguish it from the top level moderation properties?

nick-funk replied:
Unsure what you mean here. The other results also do this same pattern? There's:

- perspective
- wordList
- akismet
- externalModeration

You want a results inside the externalModeration type? Or a results with an externalModeration inside?

If the former, then we should add a results subtype to perspective, wordList, akismet, etc.
If the latter, then we need to move akismet, wordList, perspective, etc inside results too.

Also, on the topic of "distinguish it from the top level moderation properties", that is kind of the point of this whole PR after talking with Sam/Andrew. 

The clients that use this want to treat the external moderation phases as 1st class mod phases. We add a lot of visibility in the mod-details and even merge them all under the same "Details" in the mod cards. Even going as far as to mark the mod card "moderated by" to show <external_mod_phase> instead of "System".

I don't think treating these as less than "top level" is accurate anymore.

nick-funk replied:
Ah, after discussion, you meant `add a results subtype` to `externalModeration`. All good, it will be done!

nick-funk committed:
Add external mod metadata result field

nick-funk committed:
Up timeout on permalink view render so it stops failing CI

alinhle commented on src/core/client/stream/test/comments/permalink/permalinkView.spec.tsx:
```diff
@@ -91,8 +91,9 @@ beforeEach(() => {
+    { timeout: 30000 }
   );
   expect(within(tabPane).toJSON()).toMatchSnapshot();
   expect(await within(testRenderer.root).axe()).toHaveNoViolations();
```
You can undo this change. The flakyness is due to `axe()` for some reason. you can comment out the line with axe, or try updating axe packages.

alinhle commented on src/core/client/admin/components/ModerateCard/ExternalModerationSummaryContainer.tsx:
```diff
@@ -0,0 +1,249 @@
+    default:
+      return null;
+  }
+};
```
Could be a component :-)

alinhle replied:
Also would make `keyCounter` unnecessary.

alinhle commented on src/core/client/admin/components/ModerateCard/ExternalModerationSummaryContainer.tsx:
```diff
@@ -0,0 +1,249 @@
+                        | React.ReactPortal
+                        | null
+                        | undefined
+                    ) => (
```
Couldn't that just be:
```ts
                  {em.result.tags?.map(
                    (t) =>

```
?

alinhle commented on src/core/client/admin/components/ModerateCard/externalModeration.ts:
```diff
@@ -0,0 +1,110 @@
+  } | null;
+  readonly status: GQLCOMMENT_STATUS_RL;
+  tags: ReadonlyArray<CODED_TAG>;
+}
```
Should be a fragment, also gives you types :-)

alinhle commented on src/core/client/admin/components/ModerateCard/ModeratedByContainer.tsx:
```diff
@@ -67,23 +67,49 @@ const ModeratedByContainer: React.FunctionComponent<Props> = ({
   }, [onUsernameClicked, moderatedBy]);
 
-  if (!moderatedBy) {
+  const externallyModeratedBy = comment.revision?.metadata?.externalModeration?.filter(
```
This should be only set when `statusHistory.edges.length` is zero.

alinhle commented on src/core/client/admin/components/ModerateCard/externalModeration.ts:
```diff
@@ -0,0 +1,110 @@
+        status: GQLCOMMENT_STATUS_RL | null;
+        tags: GQLTAG_RL[];
+      };
+    }) => {
```
`m` is already typed

alinhle commented on src/core/client/admin/components/ModerateCard/externalModeration.ts:
```diff
@@ -0,0 +1,110 @@
+      // Check if the status matches the external moderation phase status
+      if (
+        status === m.result.status &&
+        m.result.status !== GQLCOMMENT_STATUS.NONE
```
That would only make sense when `statusHistory` array is empty

alinhle commented on src/core/client/admin/components/ModerateCard/externalModeration.ts:
```diff
@@ -0,0 +1,110 @@
+      // Check if the tags match the external moderation phase tags
+      if (
+        tags &&
+        m.result.tags?.every((t: GQLTAG_RL) =>
```
`t` is already typed?

alinhle commented on src/core/client/admin/components/ModerateCard/externalModeration.ts:
```diff
@@ -0,0 +1,110 @@
+  tags: ReadonlyArray<CODED_TAG>;
+}
+
+const filterValidExternalModItems = (comment?: Comment | null) => {
```
Why do we need to filter for valid external mod phases?

nick-funk committed:
Use custom fragment for externalModeration helper

nick-funk committed:
Convert getReasonMarker to ReasonMarker component

nick-funk committed:
Externally moderated by only set if status history is empty

nick-funk committed:
Commenting out permalink view axe test for now

Marked a TODO with notes for why this is so involved
and can't be easily fixed right now.

nick-funk committed:
More axe tests to temporarily disable

nick-funk committed:
Make the builds/tests stop complaining about types

nick-funk committed:
Remove externalModeration filter check

alinhle committed:
wip: Moderatedby fixes

nick-funk committed:
Fix some conditional null linting errors

nick-funk committed:
Update snapshots for conditional moderated by button

alinhle committed:
fix: missing snapshots

kodiakhq[bot] committed:
Merge branch 'develop' into feat/CORL-1761-mod-phase-record
